### ISSUE TITLE
Clarification about Sequential mode

### ISSUE THREAD

**Issue by elgatito** [arvidn/libtorrent#4408]
I'm trying to understand how sequential piece picker works in libtorrent.

For example consider this flow:
1) Adding torrent, selecting file to download.
2) User reads file from the beginning immediatelly. I see the piece is not available yet, so I set highest priority to the first piece. Pieces are downloaded from the 1st one sequentially now?
3) Now user wants to seek the file to the middle. I see the piece is not available and set the highest priority to that piece. How should piece picker work now?

What I'm trying to make:
Reads from user is coming through my code, so I know what it is reading now. So I want to make sure that next coming pieces are requested next and not randomly.

Currently I manually update pieces priority each second, so set kind of priority chain, like 7-6-6-5-5-5-5.... to pieces, and when user seeks I update priorities to new starting place.

But maybe there are much better options. Because on not-so-fast devices and torrents with big number of pieces I see it takes more time to update priorities (which is logical of course).

Thanks

**Comment by ssiloti** [comment]
Sequential mode always downloads sequentially from the beginning of the torrent. For what you describe `torrent_handle::set_piece_deadline` should work better than changing piece priorities. Setting a piece deadline of zero will ensure that piece is downloaded as quickly as possible.

**Comment by elgatito** [comment]
@ssiloti consider this pseudo-code:

`
startPiece = 10
endPiece = 60
currentReadPiece = 30

for (index in (currentReadPiece, currentReadPiece+10)) {
    t.set_piece_deadline(index, (index - currentReadPiece) * 50)
}
`

So  I need to download 50 pieces, but user currently reads piece number 30, so I need to prioritize 10 pieces next it, but make sure they are downloaded kind of sequentially. `currentReadPiece + 1` with the highest priority, so calling `set_piece_deadline` with `0` and the next one with `50` and so on.

Should I expect pieces to be downloaded in that order?

**Comment by elgatito** [comment]
@ssiloti @arvidn 

I have found this test: https://github.com/arvidn/libtorrent/blob/bdee0b563cf0c6f6bbb265752b9d80768528fd12/test/test_piece_picker.cpp#L782

It looks for me that if I set all priorities to 1, and the ones I need urgently with 2 (or any bigger than 1), then picker will request pieces in my desired order?

**Comment by arvidn** [comment]
@elgatito If you set sequential mode, piece priorities are not taken into account. There is an exception to this (and now that I look at it, it's not clear that it makes sense), pieces with *top* priority (i.e. 7) are still picked before other pieces.

There's a distinction between order pieces are *requested* an the order they are *received*. Sequential mode and piece priorities, affect the order they are requested in. The piece deadline is attempting to make the pieces *received* in the order of their deadlines.

> So I need to download 50 pieces, but user currently reads piece number 30, so I need to prioritize 10 pieces next it, but make sure they are downloaded kind of sequentially. currentReadPiece + 1 with the highest priority, so calling set_piece_deadline with 0 and the next one with 50 and so on.
> 
> Should I expect pieces to be downloaded in that order?

Yes. But in reality, pieces are downloaded in parallel and completing in somewhat random order.

**Comment by elgatito** [comment]
@arvidn I understand that there are many things that influence order of getting pieces (like number of peers, their speeds, piece availability, network, etc), but I mainly bother about order of requesting pieces. Imho, in average, pieces will be received in same order as they have been requested, or no?

> Yes. But in reality, pieces are downloaded in parallel and completing in somewhat random order.

That brings me back to setting all priorities manually, to make sure picker takes prioritized pieces first...

**Comment by arvidn** [comment]
> but I mainly bother about order of requesting pieces. Imho, in average, pieces will be received in same order as they have been requested, or no?

That's not my experience. Some peers don't respond at all to requests and some deliver data at megabytes per second. It's a wide range.

> That brings me back to setting all priorities manually, to make sure picker takes prioritized pieces first...

I think it's reasonable advice to not use sequential mode if you want to stream content.

**Comment by elgatito** [comment]
> I think it's reasonable advice to not use sequential mode if you want to stream content.

Any advice from you? Or just keep prioritizing all the pieces in a vector with set_piece_priorities?

**Comment by arvidn** [comment]
> Or just keep prioritizing all the pieces in a vector with set_piece_priorities?

Yes, either piece priorities or piece deadlines. The *intention* of `set_piece_deadline` is to be the ideal choice in this case. I've heard reports that its current implementation is CPU intensive though.

If you ever change playback position, you also probably want to cancel outstanding pieces from peers, to improve responsiveness.

**Comment by elgatito** [comment]
> Yes, either piece priorities or piece deadlines. The intention of set_piece_deadline is to be the ideal choice in this case.

With the pseudo-code implementation I could not achieve order of pieces downloads.

> If you ever change playback position, you also probably want to cancel outstanding pieces from peers, to improve responsiveness.

Cancel outstanding pieces? How can that be done? You mean calling `clear_piece_deadlines` when I have a position change and I need to reorder pieces priorities?

**Comment by arvidn** [comment]
> Cancel outstanding pieces? How can that be done? You mean calling clear_piece_deadlines when I have a position change and I need to reorder pieces priorities?

I don't recall off-hand if there is a way to ask libtorrent to cancel requests. But I expect that you'll want to at some point. Please feel free to post an issue when you're there and we can look at it. cancelling a piece deadline seems like a reasonable thing to hook that into.

**Comment by elgatito** [comment]
> I don't recall off-hand if there is a way to ask libtorrent to cancel requests.

Yes, that is, currently, the point I miss. 
When user changes read position, and libtorrent, at that moment, is fully using available bandwidth, then newly requested pieces are downloaded with slower because previously requested pieces are still coming to us (in my understanding). 
Nothing bad with that, but if I need urgently few blocks - I cannot maximize download speed for them, currently, right?

**Comment by arvidn** [comment]
It's been quite a while since I was working on streaming in libtorrent so I don't have all the trade-offs fresh in my mind. But I'm pretty optimistic that it's possible to get good throughput and reasonably responsive change of read position. Cancelling outstanding blocks that are no longer high priority, and making sure the request queue depth doesn't grow out of control is important, I think.

**Comment by elgatito** [comment]
> Cancelling outstanding blocks

Currently I see this approach as optimal:
- Update pieces priority to reflect what I need in the first order, like priorities for pieces: 111766555544444411111, so first "7" will be downloaded, then all "6"s and so on.
- If seek happens, I should call `clear_piece_deadlines` (just to be sure deadlined pieces are reset),  then `set_piece_deadline` to next N pieces, to they come first, and refreshing priorities to keep proper order when deadlined pieces are downloaded. 

What will happen, if I can use 100Mbit channel, torrent is very crowded, so that usually all the bandwidth is used easily, and I set few pieces to deadlines? Will the speed be still the same? 
I remember, last time, when I tried deadlines, I saw smaller speeds. Looked like the library requested deadlined pieces, while other pieces were "on hold" or just waited till deadlined are done.

**Comment by arvidn** [comment]
you may even want to set priority 0 to pieces that are before the playback point, and too far into the future. Otherwise you may end up wasting some download capacity requesting those pieces from peers that don't have any other.

In your description you talk about priorities in the first bullet, and deadlines in the other. I think you may want to use both in both cases. (but priorities mostly to set them to 0).

What behaves best is something you'll have to measure.

**Comment by elgatito** [comment]
> you may even want to set priority 0 to pieces that are before the playback point, and too far into the future. Otherwise you may end up wasting some download capacity requesting those pieces from peers that don't have any other.

I do actually set `priority:0` to non-needed pieces, but the problem is that Kodi (Media player) reads video with multiple clients, one can be used to read audio track, one to read actually video stream. 
So it's legal there are multiple read points at the same time.

I'm now playing with `set_piece_deadline` for urgent pieces.
Here is the log file of priorities and process of downloading. 

Priorities are logged each second, in format `index:priority:progress`.
And it prints only pieces that have `priority>0`.

Interesting part here is `1132:7:0, 1133:7:5, 1134:7:0`, I initially set these 3 pieces with `set_piece_deadline`, only they have priority `7`, but you can see that for many seconds nothing is downloaded, and the speed is pretty big. 
Then suddenly these pieces are downloaded.

Can it happen that piece picker is not picking those pieces because is occupied or does not have free slots or something like that? 

```
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:49, 722:3:41, 723:2:58, 724:2:100, 725:2:26, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:86, 731:2:21, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:0, 1134:7:0, 1135:4:0, 1136:4:0, 1137:4:0, 1138:3:0, 1139:3:0, 1140:3:0, 1141:3:0, 1142:2:0, 1143:2:0, 1144:2:0, 1145:2:0, 1146:2:0, 1147:2:0, 1148:2:0, 1149:2:0, 1150:2:0, 1151:2:0, 1152:2:0, 1153:2:0, 1154:2:0, 1155:2:0, 1156:2:0, 1157:2:0, 1158:2:0, 1159:2:0, 1160:2:0, 1161:2:0, 1162:2:0, 1163:2:0, 1164:2:0, 1165:2:0, 1166:2:0, 1167:2:0, 1168:2:0, 1169:2:0, 1170:2:0, 1171:2:0, 1172:2:0, 1173:2:0, 1174:2:0, 1175:2:0, 1176:2:0, 1177:2:0, 1178:2:0, 1179:2:0, 1180:2:0, 1181:2:0]
Prioritizing pieces: 40% / 11 MB / 0 B, Con: 18/19 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:49, 722:3:55, 723:2:64, 724:2:100, 725:2:39, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:89, 731:2:22, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:4, 1134:7:0, 1135:4:0, 1136:4:0, 1137:4:0, 1138:3:0, 1139:3:0, 1140:3:0, 1141:3:0, 1142:2:0, 1143:2:0, 1144:2:0, 1145:2:0, 1146:2:0, 1147:2:0, 1148:2:0, 1149:2:0, 1150:2:0, 1151:2:0, 1152:2:0, 1153:2:0, 1154:2:0, 1155:2:0, 1156:2:0, 1157:2:0, 1158:2:0, 1159:2:0, 1160:2:0, 1161:2:0, 1162:2:0, 1163:2:0, 1164:2:0, 1165:2:0, 1166:2:0, 1167:2:0, 1168:2:0, 1169:2:0, 1170:2:0, 1171:2:0, 1172:2:0, 1173:2:0, 1174:2:0, 1175:2:0, 1176:2:0, 1177:2:0, 1178:2:0, 1179:2:0, 1180:2:0, 1181:2:0]
Prioritizing pieces: 40% / 10 MB / 0 B, Con: 18/19 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:50, 722:3:67, 723:2:76, 724:2:100, 725:2:53, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:90, 731:2:23, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:0, 1136:4:10, 1137:4:0, 1138:3:0, 1139:3:3, 1140:3:0, 1141:3:0, 1142:2:0, 1143:2:0, 1144:2:2, 1145:2:0, 1146:2:0, 1147:2:0, 1148:2:0, 1149:2:0, 1150:2:0, 1151:2:0, 1152:2:0, 1153:2:0, 1154:2:0, 1155:2:0, 1156:2:0, 1157:2:0, 1158:2:0, 1159:2:0, 1160:2:0, 1161:2:0, 1162:2:0, 1163:2:0, 1164:2:0, 1165:2:0, 1166:2:0, 1167:2:0, 1168:2:0, 1169:2:0, 1170:2:0, 1171:2:0, 1172:2:0, 1173:2:0, 1174:2:0, 1175:2:2, 1176:2:0, 1177:2:3, 1178:2:0, 1179:2:0, 1180:2:0, 1181:2:0]
Prioritizing pieces: 40% / 11 MB / 0 B, Con: 18/19 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:50, 722:3:89, 723:2:87, 724:2:100, 725:2:63, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:27, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:3, 1136:4:31, 1137:4:17, 1138:3:0, 1139:3:16, 1140:3:7, 1141:3:7, 1142:2:0, 1143:2:0, 1144:2:14, 1145:2:0, 1146:2:14, 1147:2:0, 1148:2:0, 1149:2:0, 1150:2:0, 1151:2:0, 1152:2:0, 1153:2:0, 1154:2:0, 1155:2:0, 1156:2:0, 1157:2:0, 1158:2:0, 1159:2:0, 1160:2:0, 1161:2:0, 1162:2:0, 1163:2:0, 1164:2:8, 1165:2:0, 1166:2:0, 1167:2:0, 1168:2:0, 1169:2:0, 1170:2:0, 1171:2:15, 1172:2:0, 1173:2:0, 1174:2:0, 1175:2:18, 1176:2:0, 1177:2:23, 1178:2:0, 1179:2:0, 1180:2:0, 1181:2:0]
Prioritizing pieces: 40% / 11 MB / 0 B, Con: 18/19 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:68, 722:3:100, 723:2:100, 724:2:100, 725:2:71, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:31, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:20, 1136:4:44, 1137:4:49, 1138:3:13, 1139:3:27, 1140:3:27, 1141:3:11, 1142:2:0, 1143:2:0, 1144:2:30, 1145:2:0, 1146:2:24, 1147:2:0, 1148:2:0, 1149:2:0, 1150:2:0, 1151:2:0, 1152:2:0, 1153:2:0, 1154:2:0, 1155:2:0, 1156:2:0, 1157:2:0, 1158:2:0, 1159:2:0, 1160:2:0, 1161:2:0, 1162:2:0, 1163:2:0, 1164:2:24, 1165:2:0, 1166:2:0, 1167:2:0, 1168:2:0, 1169:2:0, 1170:2:0, 1171:2:31, 1172:2:0, 1173:2:0, 1174:2:0, 1175:2:31, 1176:2:0, 1177:2:50, 1178:2:0, 1179:2:0, 1180:2:0, 1181:2:0]
Prioritizing pieces: 40% / 11 MB / 0 B, Con: 18/19 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:89, 722:3:100, 723:2:100, 724:2:100, 725:2:96, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:36, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:34, 1136:4:56, 1137:4:72, 1138:3:33, 1139:3:37, 1140:3:45, 1141:3:15, 1142:2:0, 1143:2:0, 1144:2:46, 1145:2:0, 1146:2:38, 1147:2:0, 1148:2:0, 1149:2:0, 1150:2:0, 1151:2:0, 1152:2:0, 1153:2:0, 1154:2:0, 1155:2:0, 1156:2:0, 1157:2:0, 1158:2:0, 1159:2:0, 1160:2:0, 1161:2:0, 1162:2:0, 1163:2:0, 1164:2:39, 1165:2:0, 1166:2:1, 1167:2:0, 1168:2:0, 1169:2:0, 1170:2:0, 1171:2:51, 1172:2:0, 1173:2:0, 1174:2:0, 1175:2:44, 1176:2:0, 1177:2:77, 1178:2:0, 1179:2:0, 1180:2:0, 1181:2:0]
Prioritizing pieces: 40% / 11 MB / 0 B, Con: 18/19 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:41, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:47, 1136:4:71, 1137:4:92, 1138:3:55, 1139:3:46, 1140:3:59, 1141:3:21, 1142:2:0, 1143:2:0, 1144:2:66, 1145:2:0, 1146:2:59, 1147:2:0, 1148:2:0, 1149:2:0, 1150:2:0, 1151:2:1, 1152:2:23, 1153:2:0, 1154:2:0, 1155:2:0, 1156:2:0, 1157:2:0, 1158:2:0, 1159:2:0, 1160:2:12, 1161:2:0, 1162:2:0, 1163:2:0, 1164:2:61, 1165:2:0, 1166:2:1, 1167:2:0, 1168:2:0, 1169:2:0, 1170:2:0, 1171:2:69, 1172:2:0, 1173:2:0, 1174:2:0, 1175:2:60, 1176:2:0, 1177:2:100, 1178:2:0, 1179:2:0, 1180:2:0, 1181:2:0]
Prioritizing pieces: 40% / 11 MB / 0 B, Con: 18/19 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:45, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:60, 1136:4:90, 1137:4:100, 1138:3:83, 1139:3:57, 1140:3:76, 1141:3:27, 1142:2:14, 1143:2:0, 1144:2:87, 1145:2:0, 1146:2:65, 1147:2:0, 1148:2:0, 1149:2:0, 1150:2:0, 1151:2:21, 1152:2:51, 1153:2:0, 1154:2:0, 1155:2:0, 1156:2:0, 1157:2:0, 1158:2:0, 1159:2:0, 1160:2:35, 1161:2:0, 1162:2:0, 1163:2:0, 1164:2:80, 1165:2:0, 1166:2:2, 1167:2:0, 1168:2:0, 1169:2:0, 1170:2:0, 1171:2:91, 1172:2:0, 1173:2:0, 1174:2:0, 1175:2:78, 1176:2:0, 1177:2:100, 1178:2:0, 1179:2:0, 1180:2:0, 1181:2:0]
Prioritizing pieces: 40% / 11 MB / 0 B, Con: 18/19 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:50, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:75, 1136:4:100, 1137:4:100, 1138:3:100, 1139:3:68, 1140:3:94, 1141:3:33, 1142:2:41, 1143:2:2, 1144:2:100, 1145:2:0, 1146:2:86, 1147:2:0, 1148:2:0, 1149:2:0, 1150:2:0, 1151:2:43, 1152:2:72, 1153:2:0, 1154:2:0, 1155:2:7, 1156:2:0, 1157:2:1, 1158:2:0, 1159:2:0, 1160:2:59, 1161:2:0, 1162:2:0, 1163:2:10, 1164:2:100, 1165:2:0, 1166:2:2, 1167:2:0, 1168:2:0, 1169:2:0, 1170:2:0, 1171:2:100, 1172:2:0, 1173:2:0, 1174:2:0, 1175:2:98, 1176:2:0, 1177:2:100, 1178:2:0, 1179:2:0, 1180:2:0, 1181:2:2]
Prioritizing pieces: 40% / 11 MB / 0 B, Con: 18/19 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:54, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:90, 1136:4:100, 1137:4:100, 1138:3:100, 1139:3:81, 1140:3:100, 1141:3:39, 1142:2:72, 1143:2:27, 1144:2:100, 1145:2:0, 1146:2:100, 1147:2:0, 1148:2:0, 1149:2:0, 1150:2:0, 1151:2:69, 1152:2:93, 1153:2:0, 1154:2:0, 1155:2:22, 1156:2:0, 1157:2:3, 1158:2:0, 1159:2:0, 1160:2:77, 1161:2:14, 1162:2:0, 1163:2:34, 1164:2:100, 1165:2:0, 1166:2:2, 1167:2:0, 1168:2:0, 1169:2:22, 1170:2:0, 1171:2:100, 1172:2:0, 1173:2:0, 1174:2:0, 1175:2:100, 1176:2:0, 1177:2:100, 1178:2:0, 1179:2:0, 1180:2:2, 1181:2:15]
Prioritizing pieces: 40% / 11 MB / 0 B, Con: 18/19 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:59, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:100, 1136:4:100, 1137:4:100, 1138:3:100, 1139:3:94, 1140:3:100, 1141:3:44, 1142:2:96, 1143:2:49, 1144:2:100, 1145:2:0, 1146:2:100, 1147:2:0, 1148:2:0, 1149:2:0, 1150:2:0, 1151:2:92, 1152:2:100, 1153:2:5, 1154:2:0, 1155:2:39, 1156:2:0, 1157:2:26, 1158:2:0, 1159:2:0, 1160:2:96, 1161:2:33, 1162:2:0, 1163:2:59, 1164:2:100, 1165:2:0, 1166:2:3, 1167:2:0, 1168:2:0, 1169:2:46, 1170:2:0, 1171:2:100, 1172:2:0, 1173:2:0, 1174:2:19, 1175:2:100, 1176:2:0, 1177:2:100, 1178:2:0, 1179:2:0, 1180:2:17, 1181:2:26]
Prioritizing pieces: 40% / 11 MB / 0 B, Con: 18/19 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:64, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:100, 1136:4:100, 1137:4:100, 1138:3:100, 1139:3:100, 1140:3:100, 1141:3:51, 1142:2:100, 1143:2:66, 1144:2:100, 1145:2:0, 1146:2:100, 1147:2:0, 1148:2:0, 1149:2:10, 1150:2:0, 1151:2:100, 1152:2:100, 1153:2:23, 1154:2:0, 1155:2:49, 1156:2:15, 1157:2:46, 1158:2:0, 1159:2:0, 1160:2:100, 1161:2:52, 1162:2:16, 1163:2:83, 1164:2:100, 1165:2:0, 1166:2:3, 1167:2:0, 1168:2:0, 1169:2:67, 1170:2:0, 1171:2:100, 1172:2:0, 1173:2:0, 1174:2:43, 1175:2:100, 1176:2:10, 1177:2:100, 1178:2:0, 1179:2:0, 1180:2:38, 1181:2:42]
Prioritizing pieces: 40% / 11 MB / 0 B, Con: 18/19 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:69, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:100, 1136:4:100, 1137:4:100, 1138:3:100, 1139:3:100, 1140:3:100, 1141:3:57, 1142:2:100, 1143:2:89, 1144:2:100, 1145:2:0, 1146:2:100, 1147:2:0, 1148:2:0, 1149:2:24, 1150:2:0, 1151:2:100, 1152:2:100, 1153:2:38, 1154:2:0, 1155:2:66, 1156:2:32, 1157:2:54, 1158:2:0, 1159:2:0, 1160:2:100, 1161:2:72, 1162:2:38, 1163:2:99, 1164:2:100, 1165:2:0, 1166:2:4, 1167:2:0, 1168:2:0, 1169:2:75, 1170:2:0, 1171:2:100, 1172:2:0, 1173:2:0, 1174:2:59, 1175:2:100, 1176:2:28, 1177:2:100, 1178:2:0, 1179:2:0, 1180:2:55, 1181:2:58]
Prioritizing pieces: 40% / 11 MB / 0 B, Con: 18/19 + 1/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:73, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:100, 1136:4:100, 1137:4:100, 1138:3:100, 1139:3:100, 1140:3:100, 1141:3:65, 1142:2:100, 1143:2:100, 1144:2:100, 1145:2:0, 1146:2:100, 1147:2:0, 1148:2:0, 1149:2:41, 1150:2:13, 1151:2:100, 1152:2:100, 1153:2:55, 1154:2:0, 1155:2:86, 1156:2:53, 1157:2:57, 1158:2:0, 1159:2:18, 1160:2:100, 1161:2:92, 1162:2:65, 1163:2:100, 1164:2:100, 1165:2:0, 1166:2:4, 1167:2:0, 1168:2:0, 1169:2:86, 1170:2:0, 1171:2:100, 1172:2:0, 1173:2:0, 1174:2:83, 1175:2:100, 1176:2:51, 1177:2:100, 1178:2:0, 1179:2:0, 1180:2:81, 1181:2:70]
Prioritizing pieces: 40% / 11 MB / 0 B, Con: 19/20 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:78, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:100, 1136:4:100, 1137:4:100, 1138:3:100, 1139:3:100, 1140:3:100, 1141:3:70, 1142:2:100, 1143:2:100, 1144:2:100, 1145:2:0, 1146:2:100, 1147:2:0, 1148:2:13, 1149:2:52, 1150:2:31, 1151:2:100, 1152:2:100, 1153:2:68, 1154:2:0, 1155:2:100, 1156:2:71, 1157:2:57, 1158:2:0, 1159:2:34, 1160:2:100, 1161:2:100, 1162:2:84, 1163:2:100, 1164:2:100, 1165:2:0, 1166:2:5, 1167:2:4, 1168:2:4, 1169:2:98, 1170:2:1, 1171:2:100, 1172:2:0, 1173:2:0, 1174:2:100, 1175:2:100, 1176:2:68, 1177:2:100, 1178:2:0, 1179:2:0, 1180:2:100, 1181:2:80]
Prioritizing pieces: 40% / 10 MB / 0 B, Con: 19/20 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:83, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:100, 1136:4:100, 1137:4:100, 1138:3:100, 1139:3:100, 1140:3:100, 1141:3:78, 1142:2:100, 1143:2:100, 1144:2:100, 1145:2:0, 1146:2:100, 1147:2:0, 1148:2:35, 1149:2:61, 1150:2:45, 1151:2:100, 1152:2:100, 1153:2:81, 1154:2:0, 1155:2:100, 1156:2:82, 1157:2:57, 1158:2:10, 1159:2:45, 1160:2:100, 1161:2:100, 1162:2:100, 1163:2:100, 1164:2:100, 1165:2:0, 1166:2:5, 1167:2:10, 1168:2:21, 1169:2:100, 1170:2:12, 1171:2:100, 1172:2:0, 1173:2:0, 1174:2:100, 1175:2:100, 1176:2:80, 1177:2:100, 1178:2:0, 1179:2:0, 1180:2:100, 1181:2:93]
Prioritizing pieces: 40% / 9.9 MB / 0 B, Con: 19/20 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:88, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:100, 1136:4:100, 1137:4:100, 1138:3:100, 1139:3:100, 1140:3:100, 1141:3:86, 1142:2:100, 1143:2:100, 1144:2:100, 1145:2:0, 1146:2:100, 1147:2:0, 1148:2:54, 1149:2:71, 1150:2:57, 1151:2:100, 1152:2:100, 1153:2:92, 1154:2:20, 1155:2:100, 1156:2:100, 1157:2:57, 1158:2:19, 1159:2:59, 1160:2:100, 1161:2:100, 1162:2:100, 1163:2:100, 1164:2:100, 1165:2:3, 1166:2:6, 1167:2:19, 1168:2:39, 1169:2:100, 1170:2:24, 1171:2:100, 1172:2:0, 1173:2:0, 1174:2:100, 1175:2:100, 1176:2:95, 1177:2:100, 1178:2:0, 1179:2:5, 1180:2:100, 1181:2:100]
Prioritizing pieces: 40% / 9.8 MB / 0 B, Con: 19/20 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:92, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:100, 1136:4:100, 1137:4:100, 1138:3:100, 1139:3:100, 1140:3:100, 1141:3:94, 1142:2:100, 1143:2:100, 1144:2:100, 1145:2:0, 1146:2:100, 1147:2:13, 1148:2:75, 1149:2:82, 1150:2:73, 1151:2:100, 1152:2:100, 1153:2:100, 1154:2:42, 1155:2:100, 1156:2:100, 1157:2:59, 1158:2:29, 1159:2:76, 1160:2:100, 1161:2:100, 1162:2:100, 1163:2:100, 1164:2:100, 1165:2:25, 1166:2:9, 1167:2:34, 1168:2:62, 1169:2:100, 1170:2:36, 1171:2:100, 1172:2:4, 1173:2:0, 1174:2:100, 1175:2:100, 1176:2:100, 1177:2:100, 1178:2:0, 1179:2:14, 1180:2:100, 1181:2:100]
Prioritizing pieces: 40% / 9.7 MB / 0 B, Con: 19/20 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:97, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:0, 1133:7:5, 1134:7:0, 1135:4:100, 1136:4:100, 1137:4:100, 1138:3:100, 1139:3:100, 1140:3:100, 1141:3:100, 1142:2:100, 1143:2:100, 1144:2:100, 1145:2:0, 1146:2:100, 1147:2:34, 1148:2:92, 1149:2:94, 1150:2:83, 1151:2:100, 1152:2:100, 1153:2:100, 1154:2:69, 1155:2:100, 1156:2:100, 1157:2:63, 1158:2:42, 1159:2:96, 1160:2:100, 1161:2:100, 1162:2:100, 1163:2:100, 1164:2:100, 1165:2:47, 1166:2:15, 1167:2:39, 1168:2:83, 1169:2:100, 1170:2:53, 1171:2:100, 1172:2:18, 1173:2:0, 1174:2:100, 1175:2:100, 1176:2:100, 1177:2:100, 1178:2:0, 1179:2:25, 1180:2:100, 1181:2:100]
Prioritizing pieces: 40% / 9.8 MB / 0 B, Con: 19/20 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:100, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:9, 1133:7:15, 1134:7:11, 1135:4:100, 1136:4:100, 1137:4:100, 1138:3:100, 1139:3:100, 1140:3:100, 1141:3:100, 1142:2:100, 1143:2:100, 1144:2:100, 1145:2:0, 1146:2:100, 1147:2:50, 1148:2:100, 1149:2:100, 1150:2:92, 1151:2:100, 1152:2:100, 1153:2:100, 1154:2:89, 1155:2:100, 1156:2:100, 1157:2:70, 1158:2:53, 1159:2:100, 1160:2:100, 1161:2:100, 1162:2:100, 1163:2:100, 1164:2:100, 1165:2:68, 1166:2:26, 1167:2:54, 1168:2:100, 1169:2:100, 1170:2:67, 1171:2:100, 1172:2:31, 1173:2:7, 1174:2:100, 1175:2:100, 1176:2:100, 1177:2:100, 1178:2:13, 1179:2:38, 1180:2:100, 1181:2:100]
Prioritizing pieces: 40% / 9.7 MB / 0 B, Con: 19/20 + 0/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:100, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:19, 1133:7:33, 1134:7:45, 1135:4:100, 1136:4:100, 1137:4:100, 1138:3:100, 1139:3:100, 1140:3:100, 1141:3:100, 1142:2:100, 1143:2:100, 1144:2:100, 1145:2:1, 1146:2:100, 1147:2:66, 1148:2:100, 1149:2:100, 1150:2:100, 1151:2:100, 1152:2:100, 1153:2:100, 1154:2:100, 1155:2:100, 1156:2:100, 1157:2:73, 1158:2:62, 1159:2:100, 1160:2:100, 1161:2:100, 1162:2:100, 1163:2:100, 1164:2:100, 1165:2:82, 1166:2:44, 1167:2:70, 1168:2:100, 1169:2:100, 1170:2:79, 1171:2:100, 1172:2:44, 1173:2:19, 1174:2:100, 1175:2:100, 1176:2:100, 1177:2:100, 1178:2:27, 1179:2:55, 1180:2:100, 1181:2:100]
Prioritizing pieces: 40% / 9.8 MB / 0 B, Con: 19/20 + 2/75
Priorities: [713:7:100, 714:5:100, 715:5:100, 716:4:100, 717:4:100, 718:4:100, 719:3:100, 720:3:100, 721:3:100, 722:3:100, 723:2:100, 724:2:100, 725:2:100, 726:2:100, 727:2:100, 728:2:100, 729:2:100, 730:2:100, 731:2:100, 732:2:100, 733:2:100, 734:2:100, 735:2:100, 736:2:100, 737:2:100]
[1132:7:25, 1133:7:52, 1134:7:85, 1135:4:100, 1136:4:100, 1137:4:100, 1138:3:100, 1139:3:100, 1140:3:100, 1141:3:100, 1142:2:100, 1143:2:100, 1144:2:100, 1145:2:10, 1146:2:100, 1147:2:83, 1148:2:100, 1149:2:100, 1150:2:100, 1151:2:100, 1152:2:100, 1153:2:100, 1154:2:100, 1155:2:100, 1156:2:100, 1157:2:80, 1158:2:74, 1159:2:100, 1160:2:100, 1161:2:100, 1162:2:100, 1163:2:100, 1164:2:100, 1165:2:100, 1166:2:73, 1167:2:89, 1168:2:100, 1169:2:100, 1170:2:90, 1171:2:100, 1172:2:59, 1173:2:32, 1174:2:100, 1175:2:100, 1176:2:100, 1177:2:100, 1178:2:43, 1179:2:77, 1180:2:100, 1181:2:100]
```

**Comment by elgatito** [comment]
Some explanations.
First we have a range of pieces which are currently read and downloaded. 
Then seek occurs and we get another range of pieces, first 3 of them are marked with priority 7 and `set_piece_deadline` with `0`, `100`, and `200`.

So you see that pieces with `set_piece_deadline` are downloaded somehow after pieces with priority `4` and `3` are downloaded.

**Comment by arvidn** [comment]
> Can it happen that piece picker is not picking those pieces because is occupied or does not have free slots or something like that?

The piece picker is not involved in picking urgent pieces, but it still runs to request pieces from "slow" peers. It's possible that you've filled the request pipeline already, and need to wait to receive blocks you've already requested before the urgent pieces are requested.

You may want to enabled `TORRENT_DEBUG_STREAMING` to 1 or 2 (those are different verbosity levels, printed to `stdout`).

Is your log printing *received* pieces or *requested* pieces?

> So you see that pieces with set_piece_deadline are downloaded somehow after pieces with priority 4 and 3 are downloaded.

If you set the priorities first, those pieces will be requested and possibly fill the pipe. If you set the deadlines after, they will end up at the end of the queue.

Right now, each peer has a limit of the number of outstanding requests, for urgent pieces, that limit is twice as high, in an attempt to have pieces with deadline take priority. Perhaps you could try lowering that limit as well. it's controlled by [request_queue_time](https://libtorrent.org/reference-Settings.html#request_queue_time).

**Comment by elgatito** [comment]
> The piece picker is not involved in picking urgent pieces, but it still runs to request pieces from "slow" peers. It's possible that you've filled the request pipeline already, and need to wait to receive blocks you've already requested before the urgent pieces are requested.

Yes, that looks like all the slots are occupied, so requests are just not going out.

> Is your log printing received pieces or requested pieces?

That is the progress, collected from `get_download_queue` with calculation of blocks.

> If you set the priorities first, those pieces will be requested and possibly fill the pipe. If you set the deadlines after, they will end up at the end of the queue.

Currently I do `piece_priority` and only then do `set_piece_deadline`.
I have tried now without `piece_priority` calls for urgent pieces. Looks better and comes quicker. But only when speed is not maximized. 

What settings can I try?

I have these now:
```
	("max_out_request_queue", 60000)
	("max_allowed_in_request_queue", 25000)
```

And I very often get this error:
`performance_alert: ....: performance warning: max outstanding piece requests reached`

**Comment by arvidn** [comment]
try lowering the request_queue_time setting. I don't think you want *deeper* queues, you want *shorter* ones to not delay urgent pieces.

**Comment by arvidn** [comment]
you should also make sure you have enough "read-ahead" of pieces with deadlines, to fill your bandwidth delay product. I think it would be reasonable to have at least 2 or 3 seconds worth of time critical pieces at any given time.

**Comment by elgatito** [comment]
> try lowering the request_queue_time setting. I don't think you want deeper queues, you want shorter ones to not delay urgent pieces.

The default is `3`? I have tried `2`. Seems better, but still not so good when full bandwidth is used.

> you should also make sure you have enough "read-ahead" of pieces with deadlines, to fill your bandwidth delay product. I think it would be reasonable to have at least 2 or 3 seconds worth of time critical pieces at any given time.

Not quite understanding what you mean...

Currently I manage proper order of downloading by managing priorities only (to have kind-of sequential mode). And usually everything is fine. 
Exception here is only when file is seeked and used gets to the piece which is not downloaded, so I set that piece with `set_piece_deadline`. 
So when seek happens - there are no critical pieces at that time, usually (in that cut of log file - for sure there were no critical pieces when I set it to 3 pieces).

> Right now, each peer has a limit of the number of outstanding requests, for urgent pieces, that limit is twice as high, in an attempt to have pieces with deadline take priority. Perhaps you could try lowering that limit as well.

What setting is for that limit and what controls it?

**Comment by arvidn** [comment]
> Not quite understanding what you mean...

You need to have enough high priority/low deadline pieces ahead of the read cursor to fill your download capacity. I think 2 or 3 seconds worth of pieces would seem reasonable. So, the number of pieces depends on how big the pieces are and the playback rate you have.

> Exception here is only when file is seeked and used gets to the piece which is not downloaded, so I set that piece with set_piece_deadline.
> So when seek happens - there are no critical pieces at that time, usually (in that cut of log file - for sure there were no critical pieces when I set it to 3 pieces).

The problem with this approach is that you're not cancelling any requests when you seek. I *think* if you use deadlines and reset the deadline, they may be cancelled. If not, that should be fixed, somehow. I'm hoping you can give me some input on how that should work (if it should work different from right now). Perhaps setting priority 0 should cancel requests, I don't think it does now.

In other words, what you do now still suffer from stuffing urgent pieces at the *end* of the pipeline.

> What setting is for that limit and what controls it?

https://libtorrent.org/reference-Settings.html#max_out_request_queue
https://libtorrent.org/reference-Settings.html#request_queue_time

you may want to lower this as well:

https://libtorrent.org/reference-Settings.html#whole_pieces_threshold

**Comment by arvidn** [comment]
in case you haven't seen it, there's also this: https://libtorrent.org/streaming.html

Please feel free to suggest improvements to the documentation!

**Comment by elgatito** [comment]
> In other words, what you do now still suffer from stuffing urgent pieces at the end of the pipeline.

Yes, that is what I also think about. 
_Maybe_ it can be useful if any new urgent piece is needed, requests from _less-needed_ pieces should be cancelled if the pipeline is full.

> Perhaps setting priority 0 should cancel requests, I don't think it does now.

Probably yes, because if I re-arrange priorities and set old pieces to _0_, they are still downloaded/received, because they are already requested.

Are there any differences between RC_1_1 and RC_1_2 in this?
To test with RC_1_2 I'd have to modify Swig bindings.

**Comment by arvidn** [comment]
I don't think there's been any advances in the streaming support between 1.1 and 1.2, no

**Comment by elgatito** [comment]
@arvidn 
If taking a look into `piece_picker` source code at https://github.com/arvidn/libtorrent/blob/RC_1_1/src/piece_picker.cpp#L2130:

```
			for (std::vector<int>::const_iterator i = m_pieces.begin();
				i != m_pieces.end() && piece_priority(*i) == priority_levels - 1; ++i)
```

It should be great to have a setter for lower sequential limit, something like:

```
torrent_handle->set_sequential_start(piece_index)
```

And a picker checker loop like:
```
...
for (std::vector<int>::const_iterator i = m_pieces.begin();
				i != m_pieces.end() && piece_priority(*i) == priority_levels - 1; ++i)
			{
				if (!is_piece_free(*i, pieces)) continue;
                                // Check that we honor lower limit of sequential range, which by default is 0
                                else if (*i  < sequential_start) continue;

....
```

This will be helpful when user seeks the file and we know where he lands - we can make sure picker takes the proper range. Later when all pieces to the end are downloaded , picker will pick the ones that are located before `sequential_start` anyway.

**Comment by master255** [comment]
@elgatito  +100 points to carma! :-)
Your solution isn't quite right.
It is necessary to add this limit in all algorithms of sequential downloading.
But you saved my time well. Thanks!
There's no need to change the priorities of the pieces at all.
This one variable controls which piece is being loaded from. + deadLines - increases the priority of the first part, which is most needed when rewinding.
I set 0 for the first 5 pieces, and then the data is downloaded sequentially in automatic mode.
When the download reaches the end you need to call:
torrentHandle.setSequentialStart(0);
and the start of the file will start downloading.

Maybe I'm wrong? Any arguments?

My code:
`if (options & sequential)
		{
			if (m_dirty) update_pieces();
			TORRENT_ASSERT(!m_dirty);

			for (auto i = m_pieces.begin();
				i != m_pieces.end() && piece_priority(*i) == top_priority; ++i)
			{
				if (!is_piece_free(*i, pieces) || (*i  < sequential_start)) continue;

				ret |= picker_log_alert::prio_sequential_pieces;

				num_blocks = add_blocks(*i, pieces
					, interesting_blocks, backup_blocks
					, backup_blocks2, num_blocks
					, prefer_contiguous_blocks, peer, suggested_pieces
					, options);
				if (num_blocks <= 0) return ret;
			}

			// in time critical mode, only pick high priority pieces
			if (!(options & time_critical_mode))
			{
				if (options & reverse)
				{
					for (piece_index_t i = prev(m_reverse_cursor); i >= m_cursor; --i)
					{
						if (!is_piece_free(i, pieces) || (i  < sequential_start)) continue;
						// we've already added high priority pieces
						if (piece_priority(i) == top_priority) continue;

						ret |= picker_log_alert::reverse_sequential;

						num_blocks = add_blocks(i, pieces
							, interesting_blocks, backup_blocks
							, backup_blocks2, num_blocks
							, prefer_contiguous_blocks, peer, suggested_pieces
							, options);
						if (num_blocks <= 0) return ret;
					}
				}
				else
				{
					for (piece_index_t i = m_cursor; i < m_reverse_cursor; ++i)
					{
						if (!is_piece_free(i, pieces) || (i  < sequential_start)) continue;
						// we've already added high priority pieces
						if (piece_priority(i) == top_priority) continue;

						ret |= picker_log_alert::sequential_pieces;

						num_blocks = add_blocks(i, pieces
							, interesting_blocks, backup_blocks
							, backup_blocks2, num_blocks
							, prefer_contiguous_blocks, peer, suggested_pieces
							, options);
						if (num_blocks <= 0) return ret;
					}
				}
			}
		}`

**Comment by master255** [comment]
@arvidn  maybe better code:

`		if (options & sequential)
		{
			if (m_dirty) update_pieces();
			TORRENT_ASSERT(!m_dirty);

//			for (auto i = m_pieces.begin();
//				i != m_pieces.end() && piece_priority(*i) == top_priority; ++i)
//			{
//				if (!is_piece_free(*i, pieces)) continue;
//
//				ret |= picker_log_alert::prio_sequential_pieces;
//
//				num_blocks = add_blocks(*i, pieces
//					, interesting_blocks, backup_blocks
//					, backup_blocks2, num_blocks
//					, prefer_contiguous_blocks, peer, suggested_pieces
//					, options);
//				if (num_blocks <= 0) return ret;
//			}

			// in time critical mode, only pick high priority pieces
			if (!(options & time_critical_mode))
			{
				if (options & reverse)
				{
					for (piece_index_t i = prev(m_reverse_cursor); i >= m_cursor; --i)
					{
						if (!is_piece_free(i, pieces)) continue;
						// we've already added high priority pieces
						if (piece_priority(i) == top_priority) continue;

						ret |= picker_log_alert::reverse_sequential;

						num_blocks = add_blocks(i, pieces
							, interesting_blocks, backup_blocks
							, backup_blocks2, num_blocks
							, prefer_contiguous_blocks, peer, suggested_pieces
							, options);
						if (num_blocks <= 0) return ret;
					}
				}
				else
				{
                    piece_index_t i = m_cursor > sequential_start ? m_cursor : sequential_start;
					for (; i < m_reverse_cursor; ++i)
					{
						if (!is_piece_free(i, pieces)) continue;
						// we've already added high priority pieces
						if (piece_priority(i) == top_priority) continue;

						ret |= picker_log_alert::sequential_pieces;

						num_blocks = add_blocks(i, pieces
							, interesting_blocks, backup_blocks
							, backup_blocks2, num_blocks
							, prefer_contiguous_blocks, peer, suggested_pieces
							, options);
						if (num_blocks <= 0) return ret;
					}
				}
			}
		}`

**Comment by arvidn** [comment]
> Without it, it is impossible to rewind files.

What does it mean to "rewind" a file?

I would like to understand why sequential mode is necessary at all. I would prefer to fix piece deadlines if they aren't sufficient over adding more hacks to sequential mode (in fact, I would prefer to phase out sequential mode. It's not an intuitive feature and a lot of people misunderstand what it is and how it works).

**Comment by master255** [comment]
@arvidn

> What does it mean to "rewind" a file?

 Rewinding a file is when a user wants to watch a movie from the middle.

> I would like to understand why sequential mode is necessary at all.

I think you forgot that there's a lot of pieces in the movie. My movie contains 10,000 (and more) pieces of 4 megabytes.
So, if I define any parameters for each piece, it will take a lot of CPU.
Therefore priorities and downloads in sequential mode should be done automatically.
I think we need to completely remove the functionality of the pieces prioritization and add the functionality that we have shown above.
If the data is to be downloaded sequentially and you can set the starting number of the piece to be downloaded sequentially, this is enough for the streaming!
Just do it!

**Comment by arvidn** [comment]
the main issue with sequential mode is that it is *not* *downloading* in sequential order, it's *requesting* in sequential order.
In my experience setting piece priorities is really cheap, and the piece picker is optimized for it. I don't see how tens of thousands of pieces can cause a lot of CPU. But maybe it is in Java.

What I believe is the CPU bottleneck is the piece-deadline implementation in libtorrent, it's not very efficient right now.

**Comment by master255** [comment]
@arvidn 

> the main issue with sequential mode is that it is not downloading in sequential order, it's requesting in sequential order.

No! It's not a problem! It's working well. If the pieces start downloading too early, the player can call "deadline (index, 0)" of the necessary piece and the whole playback will continue.
And the main thing is that it's as easy as possible for the CPU.
Deadline - are used very rarely. In general, streaming should work in automatic mode.

> In my experience setting piece priorities is really cheap, and the piece picker is optimized for it. I don't see how tens of thousands of pieces can cause a lot of CPU.

I think there's a bug.
When I set the 9999 pieces to IGNORE priority and the last piece is downloaded, the torrent download stops. After that, if I give priority to all pieces of NORMAL, the download will start only a minute later, which is a lot of time.
That's why it's impossible to use the priority pieces.

> What I believe is the CPU bottleneck is the piece-deadline implementation in libtorrent, it's not very efficient right now.

It doesn't matter. Because they are used very rarely and CPU costs are justified for them.

**Comment by master255** [comment]
@arvidn And my (and elgatito) scheme without priorities and with moving the start of sequential mode and rare use of deadlines already works. I checked it out on big 100Gb movies and not big 3Gb. It works at maximum speed.

**Comment by master255** [comment]
@arvidn when will this be implemented? And what are the reasons to use priority for pieces?

**Comment by master255** [comment]
@arvidn ?

**Comment by arvidn** [comment]
What does "this" refer to? This ticket is a question, not a bug report. I believe it has been answered

**Comment by master255** [comment]
@arvidn 1. We have come to the conclusion that we need a new function to make the sequential mode work properly:
torrent_handle.set_sequential_start(index);
This is very easy to add to any of your 1.2.10 or 2.0 branches.
2. If we implement this function, there is no need to prioritize the pieces completely. Perhaps there are some other scenarios where you need the priorities of pieces? But you don't talk about it.
Removing the pieces priorities will increase the efficiency and speed of the sequential mode.
3. I have been using this scheme for several months now and everything works fine. This feature has already been successfully tested and works perfectly in my player, which is bought for money. I have many positive feedback from my users.

What more do you need?

**Comment by arvidn** [comment]
you want to submit a patch?
if there is a start-index, it seems like there should be an end-index as well, right?
if all pieces within the range (start, end) have been picked, or none are available. Should the piece picker fall back to rarest first outside of the range?

This would need to be considered and documented

**Comment by arvidn** [comment]
> 2. If we implement this function, there is no need to prioritize the pieces completely. Perhaps there are some other scenarios where you need the priorities of pieces? But you don't talk about it.

That's right. Do you want *me* to talk about them? How about you start here: https://github.com/search?q=prioritize_pieces&type=code

> 3. I have been using this scheme for several months now and everything works fine. This feature has already been successfully tested and works perfectly in my player, which is bought for money. I have many positive feedback from my users.

How do you deal with peers that are slow and don't respond very quickly, or that can only send you bytes at low rates. Do you just accept that they stall the pipe and you have to wait to resume playback?

**Comment by elgatito** [comment]
@arvidn Not sure what @master255  wants, I'd just say about my specific case.

> if there is a start-index, it seems like there should be an end-index as well, right?
if all pieces within the range (start, end) have been picked, or none are available. Should the piece picker fall back to rarest first outside of the range?

When torrent is added for streaming, it can be that user needs it to start not from beginning. 
Currently, I manually set priorities for each piece, starting from current piece, requested by user. This works fine, piece_picker is taking those pieces, but when torrent has thousand pieces and I have to update vector and call `prioritize_pieces` it sometimes takes a while. 

Sequential mode is fine for me, it works as expected, but it tries to download pieces starting from `0`. 
And in case of media file seek - I need to adjust starting point of sequential mode.

Having ability to set `end`, not only `start` is also great.

**Comment by master255** [comment]
@arvidn 

> you want to submit a patch?

Yes. I can add a pull request.

> if there is a start-index, it seems like there should be an end-index as well, right?

No, it is not. The final index is the end of the file. There is no need to specify the end index since the file is loaded sequentially.

> if all pieces within the range (start, end) have been picked, or none are available. Should the piece picker fall back to rarest first outside of the range?

No. This is not necessary. 
Each time data comes in the PIECE_FINISHED event, the player refresh the ranges of available data. When the range reaches the end of the file, the player resets the start of sequential loading:
torrent_handle.set_sequential_start(0);
And the library downloads the missing data.

> That's right. Do you want me to talk about them? How about you start here: https://github.com/search?q=prioritize_pieces&type=code

Perhaps we should start with an annotation, so that we can optionally disable the piece priority functionality during compilation.

**Comment by master255** [comment]
@arvidn End of file check is a recursive function that checks all data forwards. Of course.

**Comment by arvidn** [comment]
> No, it is not. The final index is the end of the file. There is no need to specify the end index since the file is loaded sequentially.

The last piece is the end of the *last file*, not necessarily the file you're interested in downloading sequentially.

> Each time data comes in the PIECE_FINISHED event, the player refresh the ranges of available data. When the range reaches the end of the file, the player resets the start of sequential loading:
> torrent_handle.set_sequential_start(0);

That still leaves a (brief) window of no sequential pieces being "pickable". There are a few different options for how to pick. One simple strategy, if there's no sequential-end index would be to simply wrap, and have piece 0 be sequentially after the last piece.

Not making a decision isn't really an option. Maybe you imply that nothing should be pickable/downloaded at that point. The risk then is that peer connections may be deemed redundant (if we're not interested in anything from those peers) and disconnected.

> End of file check is a recursive function that checks all data forwards. Of course.

I don't know what you mean by that.

> > How do you deal with peers that are slow and don't respond very quickly, or that can only send you bytes at low rates.

> I do not know what you are talking about. I don't have such cases.

I mean the case where you *request* pieces in order, say 0, 1, 2, 3, 4, 5. But one of the blocks of piece 2 was requested from a peer that's sending very slow to you. So the pieces *complete* in this order: 0, 1, 3, 4, 5, and you're still waiting for 2. In this scenario, you can't play beyond piece 2, unless you skip some data.

You say you don't have that case, how come? Do you disconnect slow peers very aggressively? Do you have some other reliable source to get such pieces from? It sounds like maybe you don't stream files at all, you download them fully first, and then play them.

> If there is no data from torrents, the player waits 30 seconds and plays the next file.

What if there *is* data, just not all of it? Say, piece 2 is missing but you have everything else.
If you're actually streaming, you don't know that piece 2 will be missing until *after* you've started playback, right?

**Comment by master255** [comment]
@arvidn 

> The last piece is the end of the last file, not necessarily the file you're interested in downloading sequentially.

Ok. Let the next files be loaded as usual. My player gives priority to "IGNORE" files so that they are not loaded. If no priority is given, the files must be loaded.

> You say you don't have that case, how come? Do you disconnect slow peers very aggressively?

Dear Arvidn, your library does it. I do not use anything special.

> That still leaves a (brief) window of no sequential pieces being "pickable"
So the pieces complete in this order: 0, 1, 3, 4, 5, and you're still waiting for 2.

After 5 seconds of missing data, I set 8 deadlines:
...
torrentHandle.setPieceDeadline(2, deadline);
...
And player waits 25 seconds for the data from the library. Seconds can be increased or decreased. It is possible to pause the player. There is no problem here, because if the data is slow, the user sees it and understands that he will not be able to watch the movie. Or they don't see it and watch the movie (most frequently used scenario).

>  It sounds like maybe you don't stream files at all

The sound is not correct.

> What if there is data, just not all of it? Say, piece 2 is missing but you have everything else.
If you're actually streaming, you don't know that piece 2 will be missing until after you've started playback, right?

In practice, this is not the case. In the management of downloads the user sees whether the download is going or not.
Problems with missed data are solved by setting deadlines.

**Comment by master255** [comment]
@arvidn Did you get my email?

**Comment by master255** [comment]
@arvidn Do you want my dollar? Or what?

**Comment by master255** [comment]
@arvidn Will you ever answer?

**Comment by markmdscott** [comment]
Hey @arvidn I'm +1 on @master255 's proposed patch as I think this is an important use case to support. Can you please move on this and allow progress to be made here re a pull request?

**Comment by arvidn** [comment]
Patches are always welcome. I don't know of anything blocked on me for this.
Things to keep in mind:

For patches against branches that have already been released, ABI needs to be preserved. That means adding flags to flag fields with extra space is OK, and adding new functions is OK.

Changes to the piece picker also needs additional tests to cover new features.

**Comment by markmdscott** [comment]
Oh okay! From the comments, I thought you weren't answering Oleg's emails for some reason...

**Comment by master255** [comment]
@markmdscott Yes. That's right. Arvidn doesn't answer my letters. That's why I decided not to do anything. Maybe he doesn't like my ideas or is afraid of them.
My super player has long been using this new technology to show any movie from torrents online.
To be honest - I don't know if anyone else can make such a good player as mine. I haven't seen any competitors for a long time, even from afar.
Well, I'll try to do a pull request, soon...
The way I see it - we need to fix this bug. It would be a big game changer.

**Comment by arvidn** [comment]
> Yes. That's right. Arvidn doesn't answer my letters.

You mean where you asked me to download your app? It wasn't clear you wanted a reply to that one.

>  That's why I decided not to do anything. Maybe he doesn't like my ideas or is afraid of them.

Patches are welcome, they are submitted on github.

**Comment by master255** [comment]
@arvidn I traced that you never installed my player. If you did, you should answer, I did.
Or it is customary to say "thank you". So I know everything's okay.

**Comment by master255** [comment]
@arvidn My pool reqest. Just try not to approve! https://github.com/arvidn/libtorrent/pull/5887

**Comment by markmdscott** [comment]
Hi @arvidn , @master255 It's great that you're finally able to reconnect and work towards this PR to improve streaming in libtorrent. I think perhaps there was a misunderstanding earlier and maybe things got a little too heated as a result. Hopefully we're on the mend. If I may be of help please let me know!

**Comment by elgatito** [comment]
Closing this issue. Initially it is a clarification issue.

**Comment by markmdscott** [comment]
@elgatito given that this issue was the impetus for the current PR (#5996) for facilitating enhanced sequential downloads, do you have comments on the current direction of this PR, suggestions for improvements in terms of the API and/or implementation?
